Add STAC <> Zarr report (#143)
cloudnativegeo/cloud-optimized-geospatial-formats-guide: cookbooks/zarr-stac-report/data-producers/index.qmd

---
title: "Zarr + STAC for Data Producers"
subtitle: "Approaches to structuring your data cubes using Zarr + STAC"
---

This section is targeted at people who are trying to disseminate data cubes. It addresses questions like:

 - Should you use STAC?
 - Should you use Zarr?
 - Where should Zarr stores go in your STAC catalog?
 - How much metadata should you pull up out of the Zarr dataset?
 - How to pull metadata out of Zarr?

> ## Is your data spatial-temporal?

Reviewer: This content interesting but I think it is a safe assumption that people are here to disseminate geospatial data and we may just want to highlight some of the benefits of STAC in this section, like it's wide adoption, rich ecosystem and integrations, interoperability with tools like QGIS and extensibility.

Author: Yeah. I learned about a lot of different catalogs during the FROST conversations a few months ago, but I agree it's not the right spot to dump that.

> ## Is your data spatial-temporal?

Reviewer: ```suggestion
## Should you use STAC?
```

IMO this is easier to follow if the headings follow your opening questions.

Since this is explicitly a geospatial guide, your data is probably spatial-temporal, but just in case it is not - there is a really interesting discussion going on about a general catalog that would cover a larger set of things (read more: [FROST][1]). 

> If you are looking for something that exists now:

Reviewer 2: something that does what?

Author: I moved things around a bit to make it clear that this section is about whether you should use STAC

> If you are looking for something that exists now:

Reviewer: ```suggestion
If you are looking for examples of non-spatio-temporal catalogs that exist now:
```

- [Google dataset search][2] which takes advantage of https://schema.org/ (see also [this guide][3] describing how to use schema.org for science)
- [Registry of Open Data on AWS][4] which lets you catalog publicly-accessible datasets stored in AWS.
- [zenodo][5] which adds a little dataset-level metadata building off the concept of DOIs.
- [intake][6] which is a Python-specific catalog supporting arbitrary metadata.

> ## Is your data well-gridded?

Reviewer: ```suggestion
## Should you use Zarr?
```

I'm still waffling on this suggestion, especially since the explanation about using COGs+STAC for unaligned data is a well-suited, but I feel like some of this content fits better into https://github.com/cloudnativegeo/cloud-optimized-geospatial-formats-guide/blob/staging/zarr/intro.qmd, which we could reference as a pre-read and then say "Given you have decided Zarr is the best fit for your use case..."

Author: Yeah there is definitely some COG vs Zarr stuff sneaking in here. I can see that just getting its own page or even report. That being said, I don't think it's the worst thing to repeat ourselves a little bit and it feels really important to capture these ideas. So I feel like I would be ok just keeping this here and then expanding it into its own page later on.

Before we really get into it, it is helpful to consider the data that you are trying to disseminate. When we say data cubes we tend to mean well-aligned high-dimensional data. If your data is not well-aligned -- for instance [Level 1 or Level 2][7] satellite imagery, then you might not want to use Zarr at all.

In the case of un-aligned data, you are likely well-served by using single-band COGs stored as STAC assets on STAC items where each item represents one scene. This setup is well-understood and well-supported by existing technology. Additionally the ability to virtually point to remote chunks on disk (using a reference file spec like kerchunk or data store spec like icechunk) enables Zarr-like access regardless of the actual file format. If you _do_ want to use Zarr then you should use the **Many smaller Zarr stores** approach outlined below.

> If your data _is_ aligned as a data cube, for instance [Level 3 or Level 4][7] data, then it is well suited to Zarr and you need to make some choices about priorities. 

Reviewer: ```suggestion
If your data _is_ aligned as a data cube, for instance [Level 3 or Level 4][7] data, then it is well suited to Zarr. In order to design and implement metadata for your Zarr store and any associated STAC metadata, it is important to consider the use case priorities for your dataset.
```

## What are your highest priorities?

It's usually not possible to optimize for all benefits at the same time. That's why we recommend picking out a few of your highest priorities. Examples include:
    
1. Streamline variable-level discovery for web-based users.
2. Integrate with existing tools and workflows.
3. Enable the simplest possible access patterns for users from `<insert-programming-language>`
4. Enable web-based visualization of large-scale multi-dimensional datasets.
5. Minimize the number of GET requests for accessing large-scale multi-dimensional datasets.
6. Minimize data transfer for accessing subsets of large-scale multi-dimensional datasets.
7. Minimize the amount of infrastructure required to maintain the catalog.
8. Minimize the cost/time of generating the catalog.
9. Limit who can read from the dataset without limiting who can inspect the metadata.

For each of these examples, prioritizing one over the other impacts how much effort you dedicate to abstracting metadata into the STAC catalog. Your choices will also impact the number of GET requests needed for common access patterns and the amount of data transferred.

## Options

> So given your priorities how should you structure the division between what metadata belongs in STAC and what belongs in Zarr? To a certain extent it depends on the shape of your data. 

Reviewer: ```suggestion
Given your priorities, how should you structure the division between what metadata belongs in STAC and what belongs in Zarr? To a certain extent it depends on the shape of your data. 
```

> ### One big Zarr store

Reviewer: ```suggestion
### One big Zarr store + basic STAC collection metadata
```

If you have aligned data cubes [Level 3 and 4 data][7] or groups of data cubes that tend to be global in scale (for instance CMIP6 data or ERA5) then you will want to use this approach.

![](../../../images/one-big-zarr+stac-diagram.png)

At a minimum this setup looks like:

1. Create a Zarr store for each dataset (i.e., set of data collected using the same platform, algorithms, model, etc)
2. Define a STAC collection for each Zarr dataset.
3. In each collection include a collection-level asset containing the href link to the Zarr store.

**Pros**

- There is no metadata duplication so the STAC side is easy to maintain.
- Simple access interface for Python users - no client-side concatenation.

**Cons**

- Potentially many GETs to construct the data cube if there is no [consolidated metadata][8] file.

> - Data variables are not exposed at the STAC level, so users cannot discover relevant datasets by searching for variables.

Reviewer 2: Is there no way to list this at the Collection level?

### One big Zarr store - some STAC metadata

Same as above but with the additional step:

4. Write variable metadata about the groups up into the STAC collection as collection-level metadata using the [Data Cube Extension][9].

**Pros**

- Data variables are exposed at the STAC level, so users can discover relevant datasets by searching for variables.
- Simple access interface for Python users - no client-side concatenation.

**Cons**

> - Potentially many GETs to construct the data cube if there is no consolidated metadata file.

Reviewer: This `con` seems like a separate concern from having metadata for a Zarr store in STAC.

Author: I think I was imagining the alternative as having all the metadata in STAC and being able to fully construct the data cube with just the STAC metadata. But maybe that's not fair as a con in this scenario and should just be a pro the most metadata scenario. (This is the kind of double-counting problem I _always_ have with pros and cons list)

> - Metadata is duplicated, but it is variable-level so unlikely to change often.

Reviewer: Should we add a caveat that it will change often if it is an ongoing dataset?

Author: Will it change often? The metadata I am thinking of are things like variable long_name, units, dimensions. I don't think shape is necessary.

- Data variables are not exposed at the STAC level, so users cannot discover relevant datasets by searching for variables.

### One big Zarr store - some STAC metadata

Same as above but with the additional step:

4. Write variable metadata about the groups up into the STAC collection as collection-level metadata using the [Data Cube Extension][9].

**Pros**

- Data variables are exposed at the STAC level, so users can discover relevant datasets by searching for variables.
- Simple access interface for Python users - no client-side concatenation.

**Cons**

- Potentially many GETs to construct the data cube if there is no consolidated metadata file.
- Metadata is duplicated, but it is variable-level so unlikely to change often.

### One big Zarr store - most STAC metadata

This applies to datasets that consist of nested groups of arrays.

5. Include collection-level [Link Templates][13] for each subgroup in the Zarr store

**Pros**

- Data variables are exposed at the STAC level, so users can discover relevant datasets by searching for variables.
- Simple access interface for Python users - no client-side concatenation.

> - Access subgroups without directly without any GETs to the data store.

Author: ```suggestion
- Access subgroup directly without any GETs to the data store.
```
